privacyidea/lib/resolvers/LDAPIdResolver.py

```diff
@@ -188,28 +188,40 @@ def cache(func):
     """
     @functools.wraps(func)
     def cache_wrapper(self, *args, **kwds):
-        # If it does not exist, create the node for this instance
-        resolver_id = self.getResolverId()
-        if not resolver_id in CACHE:
-            CACHE[resolver_id] = {"getUserId": {},
-                                  "getUserInfo": {},
-                                  "_getDN": {}}
-
-        # get the portion of the cache for this very LDAP resolver
-        r_cache = CACHE.get(resolver_id).get(func.func_name)
-        if args[0] in r_cache and \
-                        datetime.datetime.now() < r_cache[args[0]][
-                    "timestamp"] + \
-                        datetime.timedelta(seconds=self.cache_timeout):
-            log.debug("Reading {0!r} from cache for {1!r}".format(args[0],
-                                                              func.func_name))
-            return r_cache[args[0]]["value"]
+        # Only run the code, in case we have a configured cache!
+        if self.cache_timeout > 0:
+            # If it does not exist, create the node for this instance
+            resolver_id = self.getResolverId()
+            now = datetime.datetime.now()
+            tdelta = datetime.timedelta(seconds=self.cache_timeout)
+            if not resolver_id in CACHE:
+                CACHE[resolver_id] = {"getUserId": {},
+                                      "getUserInfo": {},
+                                      "_getDN": {}}
+            else:
+                # Clean up the cache in the current resolver and the current function
+                _to_be_deleted = []
+                for user, cached_result in CACHE[resolver_id].get(func.func_name).iteritems():
+                    if cached_result.get("timestamp") < now + tdelta:
+                        _to_be_deleted.append(user)
+                for user in _to_be_deleted:
+                    del CACHE[resolver_id][func.func_name][user]
```

Comment:
We need to be careful with regard to thread safety here: If we run this in a multi-threaded setup, two threads could add the same key to their ``_to_be_deleted`` list, so Thread 1 will delete it from ``CACHE`` and the above line will fail in Thread 2.

Reply:
Phew, good question :-)
In this case, we *may* even get away with only wrapping this in a try-except block. However, using a lock would definitely be cleaner -- but this would also affect performance again. :-/